cpp/src/arrow/dataset/partition.h

```diff
@@ -0,0 +1,191 @@
+
+// class PartitionField {
+//  public:
+
+//  private:
+//   std::string field_name_;
+// };
+
+// ----------------------------------------------------------------------
+// Partition identifiers
+
+/// \brief A partition level identifier which can be used
+///
+/// TODO(wesm): Is this general enough? What other kinds of partition
+/// keys exist and do we need to support them?
+class PartitionKey {
+ public:
+  const std::vector<std::string>& fields() const { return fields_; }
+  const std::vector<std::shared_ptr<Scalar>>& values() const { return values_; }
+
+ private:
+  std::vector<std::string> fields_;
+  std::vector<std::shared_ptr<Scalar>> values_;
+};
+
+/// \brief Intermediate data structure for data parsed from a string
+/// partition identifier.
+///
+/// For example, the identifier "foo=5" might be parsed with a single
+/// "foo" field and the value 5. A more complex identifier might be
+/// written as "foo=5,bar=2", which would yield two fields and two
+/// values.
+///
+/// Some partition schemes may store the field names in a metadata
+/// store instead of in file paths, for example
+/// dataset_root/2009/11/... could be used when the partition fields
+/// are "year" and "month"
+struct PartitionKeyData {
+  std::vector<std::string> fields;
+  std::vector<std::shared_ptr<Scalar>> values;
+};
+
+// ----------------------------------------------------------------------
+// Partition schemes
+
+/// \brief
+class ARROW_DS_EXPORT PartitionScheme {
+ public:
+  virtual ~PartitionScheme() = default;
+
+  /// \brief The name identifying the kind of partition scheme
+  virtual std::string name() const = 0;
+
+  virtual bool PathMatchesScheme(const std::string& path) const = 0;
+
+  virtual Status ParseKey(const std::string& path, PartitionKeyData* out) const = 0;
+};
+
+/// \brief Multi-level, directory based partitioning scheme
+/// originating from Apache Hive with all data files stored in the
+/// leaf directories. Data is partitioned by static values of a
+/// particular column in the schema. Partition keys are represented in
+/// the form $key=$value in directory names
+class ARROW_DS_EXPORT HivePartitionScheme : public PartitionScheme {
+ public:
+  /// \brief Return true if path
+  bool PathMatchesScheme(const std::string& path) const override;
+
+  virtual Status ParseKey(const std::string& path, PartitionKeyData* out) const = 0;
+};
+
+// ----------------------------------------------------------------------
+//
+
+class ARROW_DS_EXPORT Partition : public DataSource {
+ public:
+  DataSource::Type type() const override;
+
+  /// \brief The key for this partition source, may be nullptr,
+  /// e.g. for the top-level partitioned source container
+  virtual const PartitionKey* key() const = 0;
+
+  virtual std::unique_ptr<DataFragmentIterator> GetFragments(
+      const Selector& selector) = 0;
+};
+
+/// \brief Container for a dataset partition, which consists of a
+/// partition identifier, subpartitions, and some data fragments
```

Comment:
I'm missing some high-level concepts here. A partition can contain _several_ fragments? Until now I assumed partition and fragment were basically similar things. But apparently they're hierarchically organized?

I think it would help if you sketched out a simple example (e.g. a directory hierarchy of CSV files) and explained to what part of the example concretely each concept maps.

Reply:
A fragment corresponds semantically to a single file, or a row group in side a Parquet file. It is an indivisible unit of data that can be scanned. 

A partition contains a collection of fragments.

I'll add an example to the comments in the header file